frontend/apps/reader/modules/readerlink.lua

```diff
@@ -663,18 +679,36 @@ function ReaderLink:onTap(_, ges)
 end
 
 --- Remember current location so we can go back to it
-function ReaderLink:addCurrentLocationToStack()
-    if self.ui.document.info.has_pages then
-        table.insert(self.location_stack, self.ui.paging:getBookLocation())
-    else
-        table.insert(self.location_stack, {
-            xpointer = self.ui.rolling:getBookLocation(),
-        })
+--- increments location_stack_index to point the the last item,
+--- and clears the next item in case there was a Forward history
+function ReaderLink:addCurrentLocationToStack(loc)
+    local location = loc
+    if location == nil then
+        if self.ui.document.info.has_pages then
+            location = self.ui.paging:getBookLocation()
+        else
+            location = {xpointer = self.ui.rolling:getBookLocation(),}
+        end
     end
+    self.location_stack_index = self.location_stack_index + 1
+    self.location_stack[self.location_stack_index] = location
+    self.location_stack[self.location_stack_index+1] = nil
 end
 
 function ReaderLink:onClearLocationStack(show_notification)
     self.location_stack = {}
+    self.location_stack_index = 0
+    if show_notification then
+        UIManager:show(Notification:new{
+            text = _("Location history cleared."),
+        })
+    end
+    return true
+end
+
+--- Clears the Forward locations by setting the next element to nil
+function ReaderLink:onClearForwardLocationStack(show_notification)
+    self.location_stack[self.location_stack_index+1] = nil
```

Comment:
Why not clear it all ?
Why keep older stalled forward locations and just clear one slot, keeping the chain broken ?

Reply:
less work :) 
but a little more complex. I am open to either way

Reply:
Is there actually anything *after* the `nil` (i.e., is this creating a sparse array)? If not, yeah, a table.remove would be clearer ;).

Reply:
Yes, things may be left after the nil, so creating a space array.

Reply:
Then it's fine with me, I assume it's an instance variable, so it'll get collected on view swap anyway ;).

Reply:
This is a sparse array to defer cleanup of forward locations until we need the space. Whenever we add an item to the end of the stack we ensure it is still the end by setting `location_stack_index+1` to nil